src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/oops/InstanceKlass.java

```diff
@@ -265,80 +272,61 @@ public long getSize() { // in number of bytes
   public static long getHeaderSize() { return headerSize; }
 
   public short getFieldAccessFlags(int index) {
-    return getFields().at(index * FIELD_SLOTS + ACCESS_FLAGS_OFFSET);
+    Field field = new Field(this, index);
```

Comment:
Anything doing a "new Field()" seems expensive. There is a lot of decompressing of the field stream that goes on each time, and no caching. I do wonder if upon creation of the InstanceKlass if we shouldn't just stream in all the fields at that point and create Field and FieldInfoValues objects for them. However, one thing I'm not sure of is to what extent we create the same InstanceKlass multiple times. If that were to be a problem, then we would have to cache InstanceKlass instances rather than just instantiate one whenever it is needed.

Reply:
Caching has been added, the stream is decoded only once, when field's metadata are requested for the first time. The whole stream is decoded and the result is cached in an array referenced by InstanceKlass.